core/src/validator.rs

```diff
@@ -1266,6 +1289,24 @@ fn new_banks_from_ledger(
         );
     }
 
+    if let Some(hard_fork_restart_slot) =
+        maybe_cluster_restart_with_hard_fork(config, bank_forks.root_bank().slot())
+    {
+        if !config.no_hard_fork_blockstore_root_reconcilation_for_local_cluster_test {
```

Comment:
I have to swallow this impurity for the completeness of integration tests (and lack of unit tests around these code). xD